protocol/msgpack/unaggregated_iterator.go

```
@@ -0,0 +1,373 @@
+
+	"gopkg.in/vmihailenco/msgpack.v2"
+)
+
+// unaggregatedIterator uses MessagePack to decode different types of unaggregated metrics.
+// It is not thread-safe.
+type unaggregatedIterator struct {
+	decoder           *msgpack.Decoder         // internal decoder that does the actual decoding
+	floatsPool        xpool.FloatsPool         // pool for float slices
+	policiesPool      pool.PoliciesPool        // pool for policies
+	metric            unaggregated.MetricUnion // current metric
+	versionedPolicies policy.VersionedPolicies // current policies
+	err               error                    // error encountered during decoding
+}
+
+// NewUnaggregatedIterator creates a new unaggregated iterator
+func NewUnaggregatedIterator(reader io.Reader, opts UnaggregatedIteratorOptions) (UnaggregatedIterator, error) {
+	if opts == nil {
+		opts = NewUnaggregatedIteratorOptions()
+	}
+	if err := opts.Validate(); err != nil {
+		return nil, err
+	}
+	it := &unaggregatedIterator{
+		decoder:      msgpack.NewDecoder(reader),
+		floatsPool:   opts.FloatsPool(),
+		policiesPool: opts.PoliciesPool(),
+	}
+
+	return it, nil
+}
+
+func (it *unaggregatedIterator) Reset(reader io.Reader) {
+	it.decoder.Reset(reader)
+	it.err = nil
+}
+
+func (it *unaggregatedIterator) Value() (*unaggregated.MetricUnion, policy.VersionedPolicies) {
+	return &it.metric, it.versionedPolicies
+}
+
+func (it *unaggregatedIterator) Err() error { return it.err }
+
+func (it *unaggregatedIterator) Next() bool {
+	if it.err != nil {
+		return false
+	}
+
+	// Resetting the metric to avoid holding onto the float64 slices
+	// in the metric field even though they may not be used
+	it.metric.Reset()
+
+	return it.decodeRootObject()
+}
+
+func (it *unaggregatedIterator) decodeRootObject() bool {
+	version := it.decodeVersion()
+	if it.err != nil {
+		return false
+	}
+	// If the actual version is higher than supported version, we skip
+	// the data for this metric and continue to the next
```

Yeah we should definitely mention that if version is higher we silently ignore incoming entries.